lib/Dialect/BaseModelica/Transforms/CallCSE.cpp

```diff
@@ -126,96 +151,283 @@ mlir::Operation *cloneDefUseChain(mlir::Operation *op,
     }
   }
 
-  mlir::IRMapping mapping;
   mlir::Operation *root = nullptr;
   for (mlir::Operation *opToClone : llvm::reverse(toClone)) {
     // Skip repeated dependencies on the same operation
-    if (mapping.contains(opToClone)) {
+    if (inductionMapping.contains(opToClone)) {
       continue;
     }
-    root = rewriter.clone(*opToClone, mapping);
+    root = rewriter.clone(*opToClone, inductionMapping);
   }
   return root;
 }
 
-void CallCSEPass::emitCse(llvm::SmallVectorImpl<CallOp> &equivalenceGroup,
-                          ModelOp modelOp, DynamicOp dynamicOp,
-                          mlir::SymbolTable &symbolTable,
-                          mlir::RewriterBase &rewriter) {
-  assert(!equivalenceGroup.empty() && "equivalenceGroup cannot be empty");
-  CallOp representative = equivalenceGroup.front();
-  const mlir::Location loc = representative.getLoc();
+/// Get the indices of induction variables accessed by [callOp].
+llvm::SmallVector<int64_t>
+getAccessedIndices(CallOp callOp, MultidimensionalRange &indices,
```

Comment:
This function needs much more documentation, as it seems to be core to the solution. It is not clear whose indices are accessed. Does "accessed indices" mean "used induction variables"? Or "accessed equation indices"? Or maybe else? The initial comment seems to suggest the latter, but the former is then what actually happens.